tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/temporal/TableauEncoder.scala

```diff
@@ -0,0 +1,423 @@
+
+  private def inBoolSet(element: TBuilderInstruction): TBuilderInstruction = builder.in(element, builder.booleanSet())
+
+  /**
+   * Encodes each of a sequence of temporal formulas.
+   * @see
+   *   [[encodeFormula]]
+   */
+  def temporalsToInvariants(modWithPreds: ModWithPreds, formulas: TlaOperDecl*): TlaModule = {
+
+    val (varDeclSeqs, predsSeq, exVarDecls) = formulas.map(singleTemporalToInvariant).unzip3
+    val varDecls = varDeclSeqs.flatten
+    val preds = predsSeq.foldLeft(PredExs())(_ ++ _)
+
+    // update init
+    val newInit = andInDecl(builder.and(preds.initExs: _*), modWithPreds.init, tracker)
+
+    // update next
+    val newNext = andInDecl(builder.and(preds.nextExs: _*), modWithPreds.next, tracker)
+
+    // update loopOK
+    val newLoopOK = andInDecl(builder.and(preds.loopOKExs: _*), modWithPreds.loopOK, tracker)
+    val loopOKopApp = builder.appOp(builder.name(newLoopOK.name, OperT1(Seq.empty, BoolT1)))
+
+    val declarationsWithUpdatedPreds = modWithPreds.module.declarations
+      .map(decl =>
+        decl.name match {
+          case newInit.name   => newInit
+          case newNext.name   => newNext
+          case newLoopOK.name => newLoopOK
+          case _              => decl
+        })
+
+    // replace each formula by formula == loopOK => formulaInitVar
+    // note: this needs to be after the loopOK predicate in the TLA module, since it depends on it
+    // this reuses the name of 'formula' so we don't have to change the invariant in the checker
+    val newFormulas = formulas.zip(exVarDecls).map { case (formula, exVarDecl) =>
+      formula.copy(body = builder.impl(loopOKopApp, builder.varDeclAsNameEx(exVarDecl)))
+    }
+
+    val declarationsWithoutFormulas = declarationsWithUpdatedPreds
+      .filterNot(decl => formulas.exists(formula => formula.name == decl.name))
+
+    // the order of declarations should be safe, but sorting doesn't hurt
+    new DeclarationSorter()(modWithPreds.module.copy(
+            declarations = varDecls ++
+              declarationsWithoutFormulas ++
+              newFormulas
+        ))
+  }
+
+  /**
+   * Encodes a given formula, using the Tableau encoding by adjusting init, next, loopOK and the given formula.
+   */
+  def singleTemporalToInvariant(formula: TlaOperDecl): (Seq[TlaVarDecl], PredExs, TlaVarDecl) = {
+
+    var (varDecls, preds, (formulaEx)) = encodeSyntaxTreeInPredicates(formula.body)
+
+    // create a new variable that stores whether the formula evaluated to true in the first state
+    // this is necessary because a temporal formula on a sequence of states should be satisfied
+    // if it holds in the first state, so we need to remember this information
+    val exVarDecl = new TlaVarDecl("__" + formula.name + "_init")(Typed(BoolT1))
+    val exVar = builder.varDeclAsNameEx(exVarDecl)
+
+    // __foo_init = [inital evaluation of foo]
+    val initExVarEx = builder.eql(
+        exVar,
+        builder.useTrustedEx(formulaEx),
```

Comment:
```suggestion
        formulaEx,
```
It's already a `TBuilderInstruction` by construction